### DISCUSSION TITLE
Countdown timer and scrolling text

### DISCUSSION THREAD

**Discussion by Fearscotty** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#628] [Q&A]
I'm looking for lots of help with setting up an 8 min countdown timer. I have no experience with this stuff so it basically needs to be done for me. this project is to help my local footy club keep time on the field.
I have the panel up and working using the below code.

This is my first time here so any information would be much appreciated.

/*************************************************************************
 * Description: 
 * 
 * The underlying implementation of the ESP32-HUB75-MatrixPanel-I2S-DMA only
 * supports output to HALF scan panels - which means outputting 
 * two lines at the same time, 16 or 32 rows apart if a 32px or 64px high panel
 * respectively. 
 * This cannot be changed at the DMA layer as it would require a messy and complex 
 * rebuild of the library's internals.
 *
 * However, it is possible to connect QUARTER (i.e. FOUR lines updated in parallel)
 * scan panels to this same library and
 * 'trick' the output to work correctly on these panels by way of adjusting the
 * pixel co-ordinates that are 'sent' to the ESP32-HUB75-MatrixPanel-I2S-DMA
 * library.
 * 
 **************************************************************************/
#include "ESP32-HUB75-MatrixPanel-I2S-DMA.h"

/* Use the Virtual Display class to re-map co-ordinates such that they draw 
 * correctly on a 32x16 1/8 Scan panel (or chain of such panels).
 */
#include "ESP32-VirtualMatrixPanel-I2S-DMA.h" 

  // Panel configuration
  #define PANEL_RES_X 64 // Number of pixels wide of each INDIVIDUAL panel module. 
  #define PANEL_RES_Y 32 // Number of pixels tall of each INDIVIDUAL panel module.
  
  
  #define NUM_ROWS 1 // Number of rows of chained INDIVIDUAL PANELS
  #define NUM_COLS 2 // Number of INDIVIDUAL PANELS per ROW
  
  // ^^^ NOTE: DEFAULT EXAMPLE SETUP IS FOR A CHAIN OF TWO x 1/8 SCAN PANELS
    
  // Change this to your needs, for details on VirtualPanel pls read the PDF!
  #define SERPENT true
  #define TOPDOWN false
  
  // placeholder for the matrix object
  MatrixPanel_I2S_DMA *dma_display = nullptr;

  // placeholder for the virtual display object
  VirtualMatrixPanel  *FourScanPanel = nullptr;
  
  /******************************************************************************
   * Setup!
   ******************************************************************************/
  void setup()

  {
  
    HUB75_I2S_CFG mxconfig(
                PANEL_RES_X*2,              // DO NOT CHANGE THIS
                PANEL_RES_Y/2,              // DO NOT CHANGE THIS
                NUM_ROWS*NUM_COLS           // DO NOT CHANGE THIS
                //,_pins            // Uncomment to enable custom pins
    );
    
mxconfig.gpio.a = 23; // led matrix pin config
mxconfig.gpio.b =19; // led matrix pin config
mxconfig.gpio.c =5; // led matrix pin config
mxconfig.gpio.d = 22; // led matrix pin config
mxconfig.gpio.e = 18; // led matrix pin config
mxconfig.gpio.clk = 21; // led matrix pin config
mxconfig.gpio.lat = 4 ; // led matrix pin config
mxconfig.gpio.oe = 15; // led matrix pin config
mxconfig.gpio.r1 = 25; // led matrix pin config
mxconfig.gpio.g1 = 26; // led matrix pin config
mxconfig.gpio.b1 = 27; // led matrix pin config
mxconfig.gpio.r2 = 14; // led matrix pin config     
mxconfig.gpio.g2 = 12; // led matrix pin config
mxconfig.gpio.b2 = 13; // led matrix pin config
    mxconfig.clkphase = false; // Change this if you see pixels showing up shifted wrongly by one column the left or right.
    
    //mxconfig.driver   = HUB75_I2S_CFG::FM6126A;     // in case that we use panels based on FM6126A chip, we can set it here before creating MatrixPanel_I2S_DMA object
  
    // OK, now we can create our matrix object
    dma_display = new MatrixPanel_I2S_DMA(mxconfig);
  
    // let's adjust default brightness to about 75%
    dma_display->setBrightness8(10);    // range is 0-255, 0 - 0%, 255 - 100%
  
    // Allocate memory and start DMA display
    if( not dma_display->begin() )
      Serial.println("****** !KABOOM! I2S memory allocation failed ***********");

   
    dma_display->clearScreen();
    delay(500);
    
    // create FourScanPanellay object based on our newly created dma_display object
    FourScanPanel = new VirtualMatrixPanel((*dma_display), NUM_ROWS, NUM_COLS, PANEL_RES_X, PANEL_RES_Y);
    
	// THE IMPORTANT BIT BELOW!
    FourScanPanel->setPhysicalPanelScanRate(FOUR_SCAN_32PX_HIGH);
  }

  
  void loop() {

      // Try again using the pixel / dma memory remapper
      int offset = PANEL_RES_X*((NUM_ROWS*NUM_COLS)-1);
      for (int i=0; i< PANEL_RES_X; i++)
      {
       //FourScanPanel->drawLine(i+offset, 0, i+offset, 7, dma_display->color565(0, 0, 255)); // blue
       
        FourScanPanel->drawLine(i+offset, 6, i+offset, 8, dma_display->color565(0, 128,0)); // g        
        FourScanPanel->drawLine(i+offset, 18, i+offset, 20, dma_display->color565(128, 0,0)); // red
        //FourScanPanel->(i+offset, 24, i+offset, 31, dma_display->color565(0, 128, 128)); // blue        
        //delay(10);
      } 

      //delay(1000);

      // Print on each chained panel 1/8 module!
      // This only really works for a single horizontal chain
      for (int i = 0; i < NUM_ROWS*NUM_COLS; i++)
      {
         
        FourScanPanel->setTextColor(FourScanPanel->color565(255, 255, 255));
        FourScanPanel->setCursor(i*PANEL_RES_X+3, FourScanPanel->height()/3); //(scott, what the fuck)
      
        
        FourScanPanel->setTextSize(1);
        FourScanPanel->print("MORE");

        FourScanPanel->setTextColor(FourScanPanel->color565(255, 255, 255));
        FourScanPanel->setCursor(i*PANEL_RES_X+30, FourScanPanel->height()/3); //(scott, what the fuck)
      
        
        FourScanPanel->print("HELP");
        
      }

      
  
  }

**Comment by mrcodetastic** [comment]
Quite a few Arduino sketches around after a Google that does this/

https://electronics-project-hub.com/arduino-countdown-timer-with-lcd-and-buzzer/

Use similar code and change the drawing code routine to use the Adafruit_GFX ->print(xxx) function instead to print the time?

**Reply by Fearscotty** [reply]
I really have no knowledge on how to put it all together.

**Comment by board707** [comment]
@Fearscotty 
What is your panel and why do you use a Four_scan_panel example for it?

**Reply by Fearscotty** [reply]
It is a 1/8 scan panel. To be honest, I don't know why. The other one I tired was doubling up on the words.

**Comment by Fearscotty** [comment]
Hi guys,

Having a read through some of the questions and answers, it looks like we can not increase the brightness. Is this correct and is there a workaround?
